### PR TITLE
FlashMintLeveraged

### PR DESCRIPTION
**Pull request by ElmoArmy**
Builds on exchangeIssuanceLeveraged 

check diff here: https://www.diffchecker.com/p2JuLunt/

Most interesting bit is the following:

a state variable `flashLoanBenefactor` is used in conjuction to a guard on receiveFlashLoan to ensure that some external account doesn't initiate a flashloan against this contract and abuse approvals.

aave flashloans didn't require this as part of the callback as those loans [forward the address of the loan's initiator](https://github.com/aave/aave-v3-core/blob/27a6d5c83560694210849d4abf09a09dec8da388/contracts/protocol/libraries/logic/FlashLoanLogic.sol#L109).

### CODE REVIEW

File: contracts/exchangeIssuance/FlashMintLeveraged.sol [hunk 1]
```diff
@@ -0,0 +1,1249 @@
...
+        debtIssuanceModule.redeem(_setToken, _amount, address(this));
+    }
+
+    /**  Returns the absolute value of x
+    * @param x The number to take the absolute value of
+    */
+    function abs(int256 x) internal pure returns (uint256) {
+        return x >= 0 ? uint256(x) : uint256(-x);
+    }
+      receive() external payable {}
```
Comment (FlattestWhite) [inline contracts/exchangeIssuance/FlashMintLeveraged.sol:1248]: nit: some weird formatting issue

File: contracts/exchangeIssuance/FlashMintLeveraged.sol [hunk 2]
```diff
@@ -0,0 +1,1249 @@
...
+     */
+    function _redeemExactSet(ISetToken _setToken, uint256 _amount) internal returns (uint256) {
+        _setToken.safeTransferFrom(msg.sender, address(this), _amount);
+        debtIssuanceModule.redeem(_setToken, _amount, address(this));
+    }
+
+    /**  Returns the absolute value of x
+    * @param x The number to take the absolute value of
+    */
+    function abs(int256 x) internal pure returns (uint256) {
```
Comment (FlattestWhite) [inline contracts/exchangeIssuance/FlashMintLeveraged.sol:1245]: Is this function used?

File: test/integration/ethereum/flashMintLeveraged.spec.ts [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+
+                beforeEach(async () => {
+                  swapDataCollateralToDebt = {
+                    path: [collateralTokenAddress, addresses.tokens.weth],
+                    fees: [500],
+                    pool: ADDRESS_ZERO,
+                    exchange: Exchange.UniV3,
+                  };
+
+                  if (inputTokenName == "collateralToken") {
```
Comment (FlattestWhite) [inline test/integration/ethereum/flashMintLeveraged.spec.ts:395]:
```suggestion
                  if (inputTokenName === "collateralToken") {
```

File: test/integration/ethereum/flashMintLeveraged.spec.ts [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+                    swapDataOutputToken = {
+                      path: [],
+                      fees: [],
+                      pool: ADDRESS_ZERO,
+                      exchange: Exchange.None,
+                    };
+                  } else {
+                    swapDataOutputToken = swapDataCollateralToDebt;
+
+                    if (inputTokenName == "WETH") {
```
Comment (FlattestWhite) [inline test/integration/ethereum/flashMintLeveraged.spec.ts:406]:
```suggestion
                    if (inputTokenName === "WETH") {
```

File: test/integration/ethereum/flashMintLeveraged.spec.ts [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+                      outputToken = weth;
+                      await weth.deposit({ value: amountIn });
+                    }
+                  }
+
+                  subjectMinAmountOut = subjectSetAmount.div(2);
+                  subjectSetToken = setToken.address;
+                  await setToken.approve(flashMintLeveraged.address, subjectSetAmount);
+
+                  if (inputTokenName != "ETH") {
```
Comment (FlattestWhite) [inline test/integration/ethereum/flashMintLeveraged.spec.ts:416]:
```suggestion
                  if (inputTokenName !== "ETH") {
```

File: test/integration/ethereum/flashMintLeveraged.spec.ts [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+                      ? await owner.wallet.getBalance()
+                      : await inputToken.balanceOf(owner.address);
+                  const inputSpent = inputBalanceBefore.sub(inputBalanceAfter);
+                  expect(inputSpent.gt(0)).to.be.true;
+                  expect(inputSpent.lte(subjectMaxAmountIn)).to.be.true;
+                });
+
+                it("should quote the correct input amount", async () => {
+                  const inputBalanceBefore =
+                    inputTokenName == "ETH"
```
Comment (FlattestWhite) [inline test/integration/ethereum/flashMintLeveraged.spec.ts:329]: small nits to make `==` into `===` and `!=` into `!==`

**Review by FlattestWhite — APPROVED** [review]

File: contracts/exchangeIssuance/FlashMintLeveraged.sol (hunk 1 above)
Comment (ElmoArmy) [reply contracts/exchangeIssuance/FlashMintLeveraged.sol]: resolved in https://github.com/IndexCoop/index-coop-smart-contracts/pull/143/commits/bbc322357d4c4e919f6378fb22e36dbb97cc75a5

File: contracts/exchangeIssuance/FlashMintLeveraged.sol (hunk 2 above)
Comment (ElmoArmy) [reply contracts/exchangeIssuance/FlashMintLeveraged.sol]: No it isn't now that we aren't doing balancer swaps in this contract.

resolved in:
https://github.com/IndexCoop/index-coop-smart-contracts/pull/143/commits/bbc322357d4c4e919f6378fb22e36dbb97cc75a5

File: test/integration/ethereum/flashMintLeveraged.spec.ts (hunk 3 above)
Comment (ElmoArmy) [reply test/integration/ethereum/flashMintLeveraged.spec.ts]: resolved in:

https://github.com/IndexCoop/index-coop-smart-contracts/pull/143/commits/15b7fac84f9f9ffcdbc1ae0ca3c90bf27af95a36

File: test/integration/ethereum/flashMintLeveraged.spec.ts (hunk 5 above)
Comment (ElmoArmy) [reply test/integration/ethereum/flashMintLeveraged.spec.ts]: resolved in:
https://github.com/IndexCoop/index-coop-smart-contracts/pull/143/commits/15b7fac84f9f9ffcdbc1ae0ca3c90bf27af95a36

File: test/integration/ethereum/flashMintLeveraged.spec.ts (hunk 6 above)
Comment (ElmoArmy) [reply test/integration/ethereum/flashMintLeveraged.spec.ts]: resolved in:
https://github.com/IndexCoop/index-coop-smart-contracts/pull/143/commits/15b7fac84f9f9ffcdbc1ae0ca3c90bf27af95a36

**Pull request merged**